### PR TITLE
Ssr implementation

### PR DESCRIPTION
**Pull request by Aharonyan**
1. SSL issue solution
2. SSR logic ready

### CODE REVIEW

File: inc/API.php [hunk 1]
```diff
@@ -30,7 +30,11 @@ public static function check_request_success($request, $function_args = [])
 
     if (is_wp_error($request)) {
       foreach ($request->errors as $notice => $message) {
-        peerboard_add_notice(sprintf('%s : %s', $notice, $message[0]), __FUNCTION__, 'error', $function_args);
+        if ($message[0] === 'cURL error 60: SSL certificate problem: certificate has expired') {
+          return 'ssl_fix';
```
Comment (sentiens) [inline inc/API.php:34]: I would also add a notice here and instruction if possible

File: inc/API.php [hunk 2]
```diff
@@ -94,8 +99,22 @@ public static function peerboard_api_call($slug, $token = 0, $body, $type = 'GET
...
+      if ($success === 'ssl_fix') {
+        // Try to update certificate and fix ssl issue 
+        $ssl_fix = self::update_wp_ca_bundle();
+
+        if (isset($ssl_fix['success'])) {
+          // Make the same request
+          $success = self::make_api_req_again($slug, $token, $body, $type, $api_url);
+          // the last solution disable ssl
+          if (!$success) {
+            self::make_api_req_again($slug, $token, $body, $type, $api_url, $check_success = true, $ssl_verify = false);
```
Comment (sentiens) [inline inc/API.php:112]: Looks like a return or request update is missing here. It's not clear why the extra "make_api_req_again" function if it just changes two args.

File: .gitignore [hunk 3]
```diff
@@ -2,3 +2,5 @@
 .idea
 
 node_modules/*
+
+composer.lock
```
Comment (sentiens) [inline .gitignore:6]: Lock files should not be excluded from sources, while vendor files should be excluded.

File: inc/SSR.php [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+
+        $useragents = require PEERBOARD_PLUGIN_DIR_PATH . '/templates/useragents.php';
+
+        foreach ($useragents as $useragent) {
+            $req_useragent = strtolower($_SERVER['HTTP_USER_AGENT']);
+            $useragent = strtolower($useragent);
+            if (strlen(strstr($req_useragent, $useragent)) > 0) {
+
+                // https://domain/post/1363374877, removing slug for some time because of bug
+                $post_url = str_replace(peerboard_get_comm_full_slug(), '', $_SERVER['REQUEST_URI']);
```
Comment (sentiens) [inline inc/SSR.php:58]: I think we should wait for core fixes and implement it right

**Review by sentiens — CHANGES_REQUESTED** [review]

File: inc/API.php (hunk 2 above)
Comment (Aharonyan) [reply inc/API.php]: I removed, additional function, it was needed for other logic but then I changed it and forgot to remove

File: inc/API.php (hunk 1 above)
Comment (Aharonyan) [reply inc/API.php:35]: Ready

File: inc/API.php [hunk 5]
```diff
@@ -92,8 +95,24 @@ public static function peerboard_api_call($slug, $token = 0, $body, $type = 'GET
...
-
-      if (!$success) {
+      // If we have issue with ssl
+      if ($success === 'ssl_fix') {
+        // Try to update certificate and fix ssl issue 
+        $ssl_fix = self::update_wp_ca_bundle();
+
+        if (isset($ssl_fix['success'])) {
+          // Make the same request
+          $success = self::peerboard_api_call($slug, $token, $body, $type, $api_url);
```
Comment (sentiens) [inline inc/API.php:105]: I think we should check $success === 'ssl_fix' here again.

Comment (sentiens) [reply inc/API.php:105]: Or we can retry some request that shouldn't be retried

File: inc/API.php [hunk 6]
```diff
@@ -90,9 +88,32 @@ public static function peerboard_api_call($slug, $token = 0, $body, $type = 'GET
...
+        // If we have issue with ssl
+        if ($message[0] === 'cURL error 60: SSL certificate problem: certificate has expired') {
+          // Try to update certificate and fix ssl issue 
+          $ssl_fix = self::update_wp_ca_bundle();
+
+          if (isset($ssl_fix['success'])) {
+            // Make the same request
+            $success = self::peerboard_api_call($slug, $token, $body, $type, $api_url);
+            // the last solution disable ssl
+            if (!$success) {
```
Comment (sentiens) [inline inc/API.php:103]: Should check ssl again

File: inc/API.php [hunk 7]
```diff
@@ -90,9 +88,32 @@ public static function peerboard_api_call($slug, $token = 0, $body, $type = 'GET
...
+              return self::peerboard_api_call($slug, $token, $body, $type, $api_url, $check_success = true, $ssl_verify = false);
```
Comment (sentiens) [inline inc/API.php:104]: Shouldn't override $check_success here and below

File: inc/API.php [hunk 8]
```diff
@@ -90,9 +88,32 @@ public static function peerboard_api_call($slug, $token = 0, $body, $type = 'GET
...
+            }
+          } else {
+            return self::peerboard_api_call($slug, $token, $body, $type, $api_url, $check_success = true, $ssl_verify = false);
+          }
+        }
+      }
+    }
+
     if ($check_success) {
       $success = self::check_request_success($request, func_get_args());
```
Comment (sentiens) [inline inc/API.php:116]: I would rename this function to check_request_success_and_report_error

File: inc/API.php (hunk 5 above)
Comment (Aharonyan) [reply inc/API.php]: Ready

File: inc/API.php (hunk 6 above)
Comment (Aharonyan) [reply inc/API.php]: Ready

**Review by sentiens — APPROVED** [review]

**Pull request merged**